core/ctree/cnode.cpp

```diff
@@ -0,0 +1,420 @@
+        for(int i = 0; i < results.num; ++i){
+            results.nodes[i]->expand(0, hidden_state_index_x, i, value_prefixs[i], policies[i]);
+            // reset
+            results.nodes[i]->is_reset = is_reset_lst[i];
+
+            cback_propagate(results.search_paths[i], min_max_stats_lst->stats_lst[i], 0, values[i], discount);
+        }
+    }
+
+    int cselect_child(CNode* root, tools::CMinMaxStats &min_max_stats, int pb_c_base, float pb_c_init, float discount, float mean_q){
+        float max_score = FLOAT_MIN;
+        const float epsilon = 0.000001;
+        std::vector<int> max_index_lst;
+        for(int a = 0; a < root->action_num; ++a){
+            CNode* child = root->get_child(a);
+            float temp_score = cucb_score(child, min_max_stats, mean_q, root->is_reset, root->visit_count - 1, root->value_prefix, pb_c_base, pb_c_init, discount);
+
+            if(max_score < temp_score){
+                max_score = temp_score;
+
+                max_index_lst.clear();
+                max_index_lst.push_back(a);
+            }
+            else if(temp_score >= max_score - epsilon){
+                max_index_lst.push_back(a);
+            }
+        }
+
+        int action = 0;
+        if(max_index_lst.size() > 0){
+            int rand_index = rand() % max_index_lst.size();
+            action = max_index_lst[rand_index];
+        }
+        return action;
+    }
+
+    float cucb_score(CNode *child, tools::CMinMaxStats &min_max_stats, float parent_mean_q, int is_reset, float total_children_visit_counts, float parent_value_prefix, float pb_c_base, float pb_c_init, float discount){
+        float pb_c = 0.0, prior_score = 0.0, value_score = 0.0;
+        pb_c = log((total_children_visit_counts + pb_c_base + 1) / pb_c_base) + pb_c_init;
+        pb_c *= (sqrt(total_children_visit_counts) / (child->visit_count + 1));
+
+        prior_score = pb_c * child->prior;
+        if (child->visit_count == 0){
+            value_score = parent_mean_q;
+        }
+        else {
+            float true_reward = child->value_prefix - parent_value_prefix;
+            if(is_reset == 1){
+                true_reward = child->value_prefix;
+            }
+            value_score = true_reward + discount * child->value();
+        }
+
+        value_score = min_max_stats.normalize(value_score);
+
+        if (value_score < 0) value_score = 0;
+        if (value_score > 1) value_score = 1;
+
+        float ucb_value = prior_score + value_score;
+        return ucb_value;
+    }
+
+    void cbatch_traverse(CRoots *roots, int pb_c_base, float pb_c_init, float discount, tools::CMinMaxStatsList *min_max_stats_lst, CSearchResults &results){
+        // set seed
```

Comment:
comments to remove. Also might be nice to add some comments in many of these methods. They are currently commentless and honestly a little difficult to read even if I know what they are supposed to be doing going into this.

Reply:
thought i already changed this but i guess not. Reinstated the code as i disabled it because of what i think is a windows specific error but i think it is still important. We need to work out if this works for windows and was just a issue on my end or not.